cpp/unreal_plugins/SimulationController/Source/SimulationController/RpcServer.h

```diff
@@ -156,14 +161,258 @@ struct FunctionWrapper<TReturn (*)(TArgs...)>
 
 } // namespace detail
 
-template <typename TFunctor>
-inline void RpcServer::bindAsync(const std::string& name, TFunctor&& functor)
+
+
+
+
+
+template <typename TFunc, typename TReturn, typename... TArgs>
+concept CCallable =
+    requires(TFunc func, TArgs... args) {
+        { func(args...) } -> std::same_as<TReturn>;
+    };
+
+template <typename TBasicEntryPointBinder>
+concept CBasicEntryPointBinder =
+    requires(TBasicEntryPointBinder basic_entry_point_binder) {
+        { basic_entry_point_binder.template bind("", []() -> void {}) } -> std::same_as<void>;
```

Comment:
There's a typo here. This should be `basic_entry_point_binder.bind()`, and not `basic_entry_point_binder.template bind`.

Reply:
I wasn't aware of this, thanks for pointing it out!
I'm a bit curious now as it's still a bit unclear to me the requirement of `template` usage here. I understand that `bind` is a dependent name, dependent on the template `TBasicEntryPointBinder`. However, I'm trying to think of a scenario where the the compiler is confused while resolving `basic_entry_point_binder.bind(...)`. I understand the ambiguity while trying to resolve something like `basic_entry_point_binder.foo<...>(...)`, but in what scenario will there be an ambiguity in resolving what appears to be a simple function call 🤔

Reply:
Ah, my bad, figured it out. I didn't know `rpc::server`'s bind function was templated, here's the definition as found in `server.h` file from rpclib library.
```c++
    template <typename F> void bind(std::string const &name, F func) {
        disp_->bind(name, func);
    }
```

Since, `bind` is a templated function, we need to specify `template` keyword when we are using it as a dependent name.

Reply:
I'm also confused. I'm currently experimenting in Compiler Explorer, and and the `.template` syntax doesn't seem to be required by any compiler when calling the `bind(...)` function inside a template. But it isn't an error either. In other words, `gcc` and `clang` and `msvc` all consider it to be optional syntax when calling a function. I'm not sure why a compiler would be more strict inside a `requires` statement than in a templated function body.

Reply:
Actually never mind. This `.template` syntax isn't even required inside the `requires` clause on Windows! I'm sure I included it for a good reason at some point, because I was not even aware of this `.template` syntax before working on the new functionality in `RpcServer.h`, so I'm sure I included it to solve a specific error I encountered. But then my code must have evolved to the point where it was no longer necessary. I just removed it and everything compiles just fine.

Reply:
I'm trying to do the same, here's my code. I'm replicating the behavior of our `rpc::server` usage here.

```c++
#include <concepts>
#include <iostream>

// class that mimics rpc::server for the purpose of testing out usage of `Dummy.foo(...)` vs `Dummy.template foo(...)` only
class Dummy
{
public:
    template <typename T>
    void foo() {
        std::cout << "Dummy::foo" << std::endl;
    }
};

template <typename T>
concept CDot = requires(T t) {
    // { t.foo() } -> std::same_as<void>; // throws an error // error: no matching function for call to 'bar<B>(B&)'
    // { t.template foo() } -> std::same_as<void>; // throws an error // error: no matching function for call to 'bar<B>(B&)'
                                                   // note: the required expression 't.foo()' is invalid
    // { t.template foo<>() } -> std::same_as<void>; // throws an error // error: no matching function for call to 'bar<B>(B&)'
                                                   // note: the required expression 't.foo()' is invalid
    { t.template foo<int>() } -> std::same_as<void>; // this works
};

void bar(CDot auto x) {
    // x.foo(); // throws an error // error: no matching function for call to 'B::foo()'
    // x.template foo(); // throws an error // error: no matching function for call to 'B::foo()'
    // x.foo<>(); // throws a warning // warning: expected 'template' keyword before dependent template name [-Wmissing-template-keyword]
                  // and throws an error // expected primary-expression before '>' token, // expected primary-expression before ')' token
    // x.template foo<>(); // throws an error // error: no matching function for call to 'B::foo<>()'
    // x.foo<int>(); // throws a warning // warning: expected 'template' keyword before dependent template name [-Wmissing-template-keyword]
                     // throws an error // error: expected primary-expression before 'int', // error: expected ';' before 'int'
    x.template foo<int>(); // this works
    std::cout << "bar" << std::endl;
}

int main()
{
    Dummy b;
    bar<Dummy>(b);
    // bar(b); // this works too

    return 0;
}
```
`gcc 13.2` throws the above errors. `clang 17.0.1`, and `msvc v19.38` also agrees with `gcc`.

Reply:
My journey in this conversation 🤣 

![image](https://github.com/isl-org/spear/assets/2341965/31af142f-5ae5-44c2-97c4-2145b962be75)

Reply:
I don't think you've exactly replicated the correct experimental conditions in your code above. The `foo` function should be templated, so that part is correct. But you should try it with and without explicitly specifying the template parameter. When calling foo without the template parameter, the compiler needs to be able to infer the template parameter. I wouldn't expect this line to work under any circumstances,

```
    // { t.foo() } -> std::same_as<void>; // throws an error // error: no matching function for call to 'bar<B>(B&)'
```

because the template parameter can't be inferred. You should try the same experiment but with the `foo` function set up to allow the compiler to automatically infer the template parameter, e.g.,

```
template <typename T>
void foo(T t) {}
```

Reply:
You are right, this works
``` c++
#include <concepts>
#include <iostream>

// class that mimics rpc::server for the purpose of testing out usage of `Dummy.foo(...)` vs `Dummy.template foo(...)` only
class Dummy
{
public:
    template <typename T>
    void foo(T t) {
        std::cout << "Dummy::foo" << std::endl;
    }
};

template <typename T>
concept CDot = requires(T t) {
    { t.foo( []()->void{} ) } -> std::same_as<void>;                  // works
    // { t.template foo( []()->void{} ) } -> std::same_as<void>;      // works
    // { t.template foo<>( []()->void{} ) } -> std::same_as<void>;    // works
};

void bar(CDot auto x) {
    x.foo( []()->void{} );          // works
    // x.template foo( []()->void{} ); // msvc doesn't like this template keyword here
    // x.foo<>( []()->void{} );        // clang doesn't like not using template keyword here
    // x.template foo<>( []()->void{} );  // works
    std::cout << "bar" << std::endl;
}

int main()
{
    Dummy b;
    bar<Dummy>(b);
    bar(b);

    return 0;
}
```

Reply:
Conclusion:
- Looks like we need to use `template` keyword whenever there is a dependent template used with angled braces (`<>`). 
    - For e.g., in [this comment](https://github.com/isl-org/spear/pull/264#discussion_r1505069798), we can see that whenever `t.foo` is used in `t.foo<>` manner, we need the `template` keyword, or else `clang` (and `gcc`) will complain, although for some reason `msvc` doesn't.

- In order to be consistent across compilers, let's use `template` keyword.

Reply:
Great summary. Our bell curve meme needs a fourth section that is even further to the right! Haha.